s3-sink-connector/src/main/java/io/aiven/kafka/connect/s3/S3SinkTask.java

```diff
@@ -115,7 +116,13 @@ public void put(final Collection<SinkRecord> records) {
     @Override
     public void flush(final Map<TopicPartition, OffsetAndMetadata> offsets) {
         try {
-            recordGrouper.records().forEach(this::flushFile);
+            final Stream<Map.Entry<String, List<SinkRecord>>> stream;
+            if (config.isWriteParallel()) {
+                stream = recordGrouper.records().entrySet().stream().parallel();
+            } else {
+                stream = recordGrouper.records().entrySet().stream();
+            }
+            stream.forEach( entry -> flushFile(entry.getKey(), entry.getValue()));
```

Comment:
Hi @aindriu-aiven happy to see how we can do this.
Can we connect and progress this together? I have production deadlines (and not using S3) 
I did design this change so that I think it should work in a common manner, not bound to a particular storage provider

As I am reading this I think that the eager writing model that I have here, is roughly what you have built for S3. Is that correct. I wanted to have the non eager model as well, so that would allow record overwriting for duplicate keys

I was also trying to ensure that what was written would work well with virtual threads & loom. I know that we can't use them yet here until we support a later JVM, but using the FJ pool and the `CompletableFuture` allows this to use this and scale later without having to rewrite continuations, and its much easier to debug!

Reply:
its does seem that the goals and the concepts overlap
I have connected to you on LinkedIn, so that we can discuss how to progress outside of a PR comment, which doesn't seem to be the correct forum